src/MDMTransaction.m

```diff
@@ -30,17 +32,58 @@ - (instancetype)init {
 }
 
 - (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target {
-  MDMTransactionLog *log = [MDMTransactionLog new];
-  log.plans = @[ [plan copy] ];
-  log.target = target;
+  [self commonAddPlan:plan toTarget:target withName:nil];
+}
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target named:(NSString *)name {
+  if (name.length && [self planForName:name withTarget:target]) {
+    [self commonRemovePlan:plan fromTarget:target named:name];
+  }
+  [self commonAddPlan:plan toTarget:target withName:name];
+}
+
+- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
+  [self commonRemovePlan:[self planForName:name withTarget:target] fromTarget:target named:name];
+}
+
+- (void)commonAddPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  MDMTransactionLog *log = [[MDMTransactionLog alloc] initWithPlan:[plan copy] target:target name:name removal:FALSE];
+  [_logs addObject:log];
+}
+
+- (void)commonRemovePlan:(NSObject<MDMPlan> *)plan fromTarget:(nonnull id)target named:(NSString *)named {
+  MDMTransactionLog *log = [[MDMTransactionLog alloc] initWithPlan:plan target:target name:named removal:TRUE];
   [_logs addObject:log];
 }
 
 - (NSArray<MDMTransactionLog *> *)logs {
   return _logs;
 }
 
+- (id<MDMPlan>)planForName:(NSString *)name withTarget:(id)target {
+  for (MDMTransactionLog *log in _logs) {
+    if ([log.name isEqualToString:name] && [log.target isEqual:target] && log.plans.count) {
+      return log.plans[0];
+    }
+  }
+  return nil;
+}
+
 @end
 
 @implementation MDMTransactionLog
+
+- (instancetype)initWithPlan:(NSObject<MDMPlan> *)plan target:(id)target name:(NSString *)name removal:(BOOL)removal {
+  self = [super init];
+  if (self) {
+    if (plan != nil) {
```

Comment:
The reason I was checking for `nil` here was because I was using this convenience initializer for creating `MDMPlans` from `commonRemovePlan` where we don't have a reference to a plan when we're trying to remove it. I guess I could use a different initializer for the remove case?

Given the fact that we want to keep the transaction object as simple as possible, I'm not sure I should be trying to look up the a plan associated with a name and a target ala `planForName:withTarget:`, hence the nil check.